File: src/facemesh_mouse/modules/cursor_image.py
"""Pure cursor bitmap / .cur-file generation -- no filesystem, no registry,
no ctypes. Builds an arrow silhouette and serializes it into the classic
(non-PNG) .cur container: ICONDIR + one ICONDIRENTRY + a combined XOR+AND
legacy DIB image, the same layout every .cur file has used since Windows
3.1. See docs/superpowers/specs/2026-08-18-cursor-appearance-design.md.
"""
from __future__ import annotations

import struct

from PIL import Image, ImageDraw

# Arrow silhouette as a fraction of the bitmap's own side length, tip at the
# origin (top-left) -- the polygon's first vertex is always the cursor's
# hotspot.
_ARROW_POINTS_FRACTION = [
    (0.0, 0.0), (0.0, 0.62), (0.18, 0.48),
    (0.29, 0.72), (0.40, 0.67), (0.29, 0.44), (0.5, 0.44),
]

VALID_MODES = {"default", "white", "black", "custom", "mista"}


def _polygon_points(size_px: int) -> list[tuple[float, float]]:
    return [(x * size_px, y * size_px) for x, y in _ARROW_POINTS_FRACTION]


def _arrow_mask(size_px: int) -> Image.Image:
    """1-channel mask, 255 inside the arrow silhouette, 0 outside."""
    mask = Image.new("L", (size_px, size_px), 0)
    ImageDraw.Draw(mask).polygon(_polygon_points(size_px), fill=255)
    return mask


def _contrast_outline_color(fill: tuple[int, int, int]) -> tuple[int, int, int]:
    """Black outline on a light fill, white outline on a dark fill --
    relative luminance (ITU-R BT.601) thresholded at the midpoint. A static
    one-time contrast choice -- "mista" mode is what protects against an
    arbitrary/changing background, this is just so a solid-color arrow
    doesn't disappear against a same-color background at the moment it's
    picked."""
    r, g, b = fill
    luminance = 0.299 * r + 0.587 * g + 0.114 * b
    return (0, 0, 0) if luminance > 127 else (255, 255, 255)


def render_color_bitmap(size_px: int, fill: tuple[int, int, int]) -> Image.Image:
    """RGBA arrow: `fill` inside, a contrasting outline traced on the
    silhouette edge, fully transparent outside."""
    mask = _arrow_mask(size_px)
    outline = _contrast_outline_color(fill)
    image = Image.new("RGBA", (size_px, size_px), (0, 0, 0, 0))
    ImageDraw.Draw(image).polygon(
        _polygon_points(size_px), fill=(*fill, 255), outline=(*outline, 255)
    )
    image.putalpha(mask)
    return image


def _pack_1bpp_rows(size_px: int, bit_is_set) -> bytes:
    """`bit_is_set(x, y) -> bool` for a size_px x size_px grid. Returns
    bottom-up rows (DIB convention), each padded to a 4-byte boundary, MSB
    of each byte = the leftmost pixel of that byte's 8-pixel span -- the
    shared row layout both the AND and XOR 1bpp masks use."""
    row_bytes = ((size_px + 31) // 32) * 4
    out = bytearray()
    for y in range(size_px - 1, -1, -1):
        row = bytearray(row_bytes)
        for x in range(size_px):
            if bit_is_set(x, y):
                row[x // 8] |= 0x80 >> (x % 8)
        out += row
    return bytes(out)


def _pack_32bpp_rows(image: Image.Image) -> bytes:
    """Bottom-up BGRA rows -- DIB byte order, not Pillow's RGBA."""
    size_px = image.width
    px = image.load()
    out = bytearray()
    for y in range(size_px - 1, -1, -1):
        for x in range(size_px):
            r, g, b, a = px[x, y]
            out += bytes((b, g, r, a))
    return bytes(out)


def _assemble_cur(
    size_px: int, bit_count: int, xor_data: bytes, and_data: bytes, hotspot: tuple[int, int]
) -> bytes:
    bmi = struct.pack(
        "<IiiHHIIiiII",
        40,  # biSize
        size_px,  # biWidth
        size_px * 2,  # biHeight -- combined XOR + AND
        1,  # biPlanes
        bit_count,  # biBitCount
        0,  # biCompression (BI_RGB)
        len(xor_data) + len(and_data),  # biSizeImage
        0, 0,  # biXPelsPerMeter, biYPelsPerMeter
        0, 0,  # biClrUsed, biClrImportant
    )
    # DIB rule: bpp<=8 requires a palette. The values are irrelevant here
    # (the mask bits, not palette indices, drive AND/XOR interpretation) --
    # only its presence is required for the header to be valid.
    color_table = bytes([0, 0, 0, 0, 255, 255, 255, 0]) if bit_count <= 8 else b""
    image_data = bmi + color_table + xor_data + and_data

    icondir = struct.pack("<HHH", 0, 2, 1)  # idType=2 -> cursor, idCount=1
    side = size_px if size_px < 256 else 0  # 0 means 256 in the ICO/CUR format
    icondirentry = struct.pack(
        "<BBBBHHII",
        side, side, 0, 0,
        hotspot[0], hotspot[1],
        len(image_data),
        6 + 16,  # offset: ICONDIR (6 bytes) + this one ICONDIRENTRY (16 bytes)
    )
    return icondir + icondirentry + image_data


def build_cur_bytes_color(image: Image.Image) -> bytes:
    size_px = image.width
    xor_data = _pack_32bpp_rows(image)
    px = image.load()
    and_data = _pack_1bpp_rows(size_px, lambda x, y: px[x, y][3] == 0)
    return _assemble_cur(size_px, bit_count=32, xor_data=xor_data, and_data=and_data, hotspot=(0, 0))


def build_cur_bytes_mista(size_px: int) -> bytes:
    """The classic invert-cursor trick: AND=1 everywhere (nothing is ever
    fully opaque-covered), XOR=1 only inside the arrow silhouette. Where
    AND=1,XOR=1 the compositor inverts the destination pixel; where
    AND=1,XOR=0 it's untouched -- so the silhouette inverts the screen
    under it and everywhere else is fully see-through, with zero ongoing
    cost to this app."""
    mask = _arrow_mask(size_px)
    mpx = mask.load()
    and_data = _pack_1bpp_rows(size_px, lambda x, y: True)
    xor_data = _pack_1bpp_rows(size_px, lambda x, y: mpx[x, y] > 127)
    return _assemble_cur(size_px, bit_count=1, xor_data=xor_data, and_data=and_data, hotspot=(0, 0))

File: apps/desktop/src/main/services/cursorImage.ts
// Pure cursor bitmap / .cur-file generation -- no filesystem, no registry,
// no native calls. Builds an arrow silhouette and serializes it into the
// classic (non-PNG) .cur container: ICONDIR + one ICONDIRENTRY + a combined
// XOR+AND legacy DIB image, the same layout every .cur file has used since
// Windows 3.1. Ported from cursor_image.py -- see
// docs/superpowers/specs/2026-08-18-cursor-appearance-design.md. No
// PIL/canvas dependency: this needs to run synchronously in the main
// process (cursor theme applies at startup, before any renderer window is
// guaranteed open), so polygon fill/outline are hand-rolled here instead of
// delegated to a renderer's 2D canvas.

export const VALID_CURSOR_MODES = new Set(["default", "white", "black", "custom", "mista"]);

// Arrow silhouette as a fraction of the bitmap's own side length, tip at the
// origin (top-left) -- the polygon's first vertex is always the cursor's
// hotspot. Identical to cursor_image.py's _ARROW_POINTS_FRACTION.
const ARROW_POINTS_FRACTION: Array<[number, number]> = [
  [0.0, 0.0], [0.0, 0.62], [0.18, 0.48],
  [0.29, 0.72], [0.4, 0.67], [0.29, 0.44], [0.5, 0.44],
];

function polygonPoints(sizePx: number): Array<[number, number]> {
  return ARROW_POINTS_FRACTION.map(([x, y]) => [x * sizePx, y * sizePx]);
}

// Point-in-polygon via the standard even-odd ray-casting test, evaluated at
// pixel centers (x+0.5, y+0.5) to match how rasterizers conventionally
// sample -- avoids the fill boundary being sensitive to exact-integer
// vertex coincidences.
function pointInPolygon(px: number, py: number, points: Array<[number, number]>): boolean {
  let inside = false;
  for (let i = 0, j = points.length - 1; i < points.length; j = i++) {
    const [xi, yi] = points[i];
    const [xj, yj] = points[j];
    const intersects = yi > py !== yj > py && px < ((xj - xi) * (py - yi)) / (yj - yi) + xi;
    if (intersects) inside = !inside;
  }
  return inside;
}

// Rasterizes the polygon's edges (not the fill) via a simple DDA line walk
// between consecutive vertices -- gives the 1px outline PIL's
// `ImageDraw.polygon(..., outline=...)` also draws, without needing a real
// stroking algorithm for this small, static 7-point shape.
function polygonEdgePixels(points: Array<[number, number]>, sizePx: number): Set<number> {
  const edgePixels = new Set<number>();
  const mark = (x: number, y: number) => {
    if (x >= 0 && x < sizePx && y >= 0 && y < sizePx) edgePixels.add(y * sizePx + x);
  };
  for (let i = 0; i < points.length; i++) {
    const [x0, y0] = points[i];
    const [x1, y1] = points[(i + 1) % points.length];
    const steps = Math.max(Math.abs(x1 - x0), Math.abs(y1 - y0), 1);
    for (let s = 0; s <= steps; s++) {
      mark(Math.round(x0 + ((x1 - x0) * s) / steps), Math.round(y0 + ((y1 - y0) * s) / steps));
    }
  }
  return edgePixels;
}

function arrowMask(sizePx: number): Uint8Array {
  // 1 where inside the arrow silhouette, 0 outside -- pixel-center sampled.
  const points = polygonPoints(sizePx);
  const mask = new Uint8Array(sizePx * sizePx);
  for (let y = 0; y < sizePx; y++) {
    for (let x = 0; x < sizePx; x++) {
      if (pointInPolygon(x + 0.5, y + 0.5, points)) mask[y * sizePx + x] = 1;
    }
  }
  return mask;
}

function contrastOutlineColor(fill: [number, number, number]): [number, number, number] {
  const [r, g, b] = fill;
  const luminance = 0.299 * r + 0.587 * g + 0.114 * b;
  return luminance > 127 ? [0, 0, 0] : [255, 255, 255];
}

// RGBA buffer (row-major, top-down, 4 bytes/pixel) -- fill inside the arrow
// silhouette, a 1px contrasting outline on the polygon edge, fully
// transparent outside.
function renderColorBitmap(sizePx: number, fill: [number, number, number]): Uint8Array {
  const mask = arrowMask(sizePx);
  const outline = contrastOutlineColor(fill);
  const edges = polygonEdgePixels(polygonPoints(sizePx), sizePx);
  const rgba = new Uint8Array(sizePx * sizePx * 4);
  for (let i = 0; i < sizePx * sizePx; i++) {
    const inside = mask[i] === 1;
    const onEdge = edges.has(i);
    const alpha = inside || onEdge ? 255 : 0;
    // Fully-transparent pixels must be RGB (0,0,0), matching the Python
    // reference's `Image.new("RGBA", ..., (0,0,0,0))` starting canvas --
    // NOT the fill/outline color. The legacy AND/XOR fallback a 32bpp
    // cursor still carries has AND=1, XOR=<this RGB> outside the shape, so
    // `dst = (dst & 1) ^ 0xFFFFFF` must reduce to a no-op (XOR 0) wherever
    // alpha is 0; XOR-ing the fill color there would invert the whole
    // cursor rectangle instead of leaving untouched pixels untouched.
    const [r, g, b] = alpha === 0 ? [0, 0, 0] : onEdge ? outline : fill;
    rgba[i * 4] = r;
    rgba[i * 4 + 1] = g;
    rgba[i * 4 + 2] = b;
    rgba[i * 4 + 3] = alpha;
  }
  return rgba;
}

function packAndMask(sizePx: number, transparentAt: (x: number, y: number) => boolean): Buffer {
  return pack1bppRows(sizePx, transparentAt);
}

// bitIsSet(x, y) for a sizePx x sizePx grid. Returns bottom-up rows (DIB
// convention), each padded to a 4-byte boundary, MSB of each byte = the
// leftmost pixel of that byte's 8-pixel span.
function pack1bppRows(sizePx: number, bitIsSet: (x: number, y: number) => boolean): Buffer {
  const rowBytes = Math.ceil(sizePx / 32) * 4;
  const out = Buffer.alloc(rowBytes * sizePx);
  for (let y = sizePx - 1, outRow = 0; y >= 0; y--, outRow++) {
    for (let x = 0; x < sizePx; x++) {
      if (bitIsSet(x, y)) {
        const byteIndex = outRow * rowBytes + (x >> 3);
        out[byteIndex] |= 0x80 >> (x % 8);
      }
    }
  }
  return out;
}

// Bottom-up BGRA rows -- DIB byte order, not the RGBA buffer's own order.
function pack32bppRows(rgba: Uint8Array, sizePx: number): Buffer {
  const out = Buffer.alloc(sizePx * sizePx * 4);
  let outOffset = 0;
  for (let y = sizePx - 1; y >= 0; y--) {
    for (let x = 0; x < sizePx; x++) {
      const i = (y * sizePx + x) * 4;
      out[outOffset++] = rgba[i + 2]; // B
      out[outOffset++] = rgba[i + 1]; // G
      out[outOffset++] = rgba[i]; // R
      out[outOffset++] = rgba[i + 3]; // A
    }
  }
  return out;
}

function assembleCur(sizePx: number, bitCount: number, xorData: Buffer, andData: Buffer): Buffer {
  const bmi = Buffer.alloc(40);
  bmi.writeInt32LE(40, 0); // biSize
  bmi.writeInt32LE(sizePx, 4); // biWidth
  bmi.writeInt32LE(sizePx * 2, 8); // biHeight -- combined XOR + AND
  bmi.writeUInt16LE(1, 12); // biPlanes
  bmi.writeUInt16LE(bitCount, 14); // biBitCount
  bmi.writeUInt32LE(0, 16); // biCompression (BI_RGB)
  bmi.writeUInt32LE(xorData.length + andData.length, 20); // biSizeImage
  // biXPelsPerMeter, biYPelsPerMeter, biClrUsed, biClrImportant all 0, already zeroed by Buffer.alloc

  // DIB rule: bpp<=8 requires a palette (values irrelevant -- mask bits, not
  // palette indices, drive AND/XOR interpretation -- only presence matters).
  const colorTable = bitCount <= 8 ? Buffer.from([0, 0, 0, 0, 255, 255, 255, 0]) : Buffer.alloc(0);
  const imageData = Buffer.concat([bmi, colorTable, xorData, andData]);

  const icondir = Buffer.alloc(6);
  icondir.writeUInt16LE(0, 0); // reserved
  icondir.writeUInt16LE(2, 2); // type: 2 = cursor
  icondir.writeUInt16LE(1, 4); // count: 1 image

  const side = sizePx < 256 ? sizePx : 0; // 0 means 256 in the ICO/CUR format
  const icondirentry = Buffer.alloc(16);
  icondirentry.writeUInt8(side, 0); // width
  icondirentry.writeUInt8(side, 1); // height
  icondirentry.writeUInt8(0, 2); // color count (0 = no palette limit)
  icondirentry.writeUInt8(0, 3); // reserved
  icondirentry.writeUInt16LE(0, 4); // xHotspot
  icondirentry.writeUInt16LE(0, 6); // yHotspot -- both 0: tip of the arrow, matching ARROW_POINTS_FRACTION's first vertex
  icondirentry.writeUInt32LE(imageData.length, 8); // bytes in image data
  icondirentry.writeUInt32LE(6 + 16, 12); // offset: ICONDIR (6 bytes) + this one ICONDIRENTRY (16 bytes)

  return Buffer.concat([icondir, icondirentry, imageData]);
}

export function buildCurBytesColor(sizePx: number, fillRgb: [number, number, number]): Buffer {
  const rgba = renderColorBitmap(sizePx, fillRgb);
  const xorData = pack32bppRows(rgba, sizePx);
  const andData = packAndMask(sizePx, (x, y) => rgba[(y * sizePx + x) * 4 + 3] === 0);
  return assembleCur(sizePx, 32, xorData, andData);
}

// The classic invert-cursor trick: AND=1 everywhere (nothing is ever fully
// opaque-covered), XOR=1 only inside the arrow silhouette. Where AND=1,XOR=1
// the compositor inverts the destination pixel; where AND=1,XOR=0 it's
// untouched -- so the silhouette inverts the screen under it and everywhere
// else is fully see-through, with zero ongoing cost to this app.
export function buildCurBytesMista(sizePx: number): Buffer {
  const mask = arrowMask(sizePx);
  const andData = pack1bppRows(sizePx, () => true);
  const xorData = pack1bppRows(sizePx, (x, y) => mask[y * sizePx + x] === 1);
  return assembleCur(sizePx, 1, xorData, andData);
}
